lib/chef/resource/openssl_x509_certificate.rb

```diff
@@ -185,24 +187,27 @@ class OpensslX509Certificate < Chef::Resource
       end
 
       action_class do
-        def generate_key_file
-          unless new_resource.key_file
-            path, file = ::File.split(new_resource.path)
-            filename = ::File.basename(file, ::File.extname(file))
-            new_resource.key_file path + "/" + filename + ".key"
-          end
-          new_resource.key_file
+        def key_file
+          @key_file ||=
+            begin
+              if new_resource.key_file
+                new_resource.key_file
+              else
+                path, file = ::File.split(new_resource.path)
+                filename = ::File.basename(file, ::File.extname(file))
+                path + "/" + filename + ".key"
+              end
+            end
```

Comment:
```suggestion
            if new_resource.key_file
              new_resource.key_file
            else
              path, file = ::File.split(new_resource.path)
              filename = ::File.basename(file, ::File.extname(file))
              path + "/" + filename + ".key"
            end
```

The `begin...end` is superfluous here.